src/lib/pubkey/sphincsplus/sphincsplus_common/sphincsplus.cpp

```diff
@@ -19,14 +20,40 @@
 #include <botan/internal/sp_xmss.h>
 #include <botan/internal/stl_util.h>
 
-#if !defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHA2) and !defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHAKE)
+#include <utility>
+
+#if !defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHA2) and !defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHAKE) and \
+   !defined(BOTAN_HAS_SLH_DSA_WITH_SHA2) and !defined(BOTAN_HAS_SLH_DSA_WITH_SHAKE)
 static_assert(
    false,
-   "botan module 'sphincsplus_common' is useful only when enabling at least 'sphincsplus_sha2' or 'sphincsplus_shake'");
+   "botan module 'sphincsplus_common' is useful only when enabling at least 'sphincsplus_sha2', 'sphincsplus_shake', 'slh_dsa_sha2', or 'slh_dsa_shake'");
 #endif
 
 namespace Botan {
 
+namespace {
+// FIPS 205, Algorithm 22, line 8
+SphincsMessageInternal prepare_message(SphincsInputMessage&& msg,
+                                       const Sphincs_Parameters& params,
+                                       StrongSpan<const SphincsContext> context) {
+   BOTAN_ARG_CHECK(params.is_slh_dsa() || context.empty(), "Context is not supported for SPHINCS+");
+   SphincsMessageInternal msg_internal{.prefix = SphincsMessagePrefix(), .message = std::move(msg)};
+   if(params.is_slh_dsa()) {
+      // prefix (no pre-hash mode): input mode byte + |ctx| + ctx
+      msg_internal.prefix = SphincsMessagePrefix(1 + 1 + context.size());
+      BufferStuffer prefix_stuffer(msg_internal.prefix);
+      const uint8_t input_mode_byte = 0x00;  // Pure (TODO: pre-hash mode: 0x01)
+      prefix_stuffer.append(input_mode_byte);
+      prefix_stuffer.append(checked_cast_to<uint8_t>(context.size()));
+      prefix_stuffer.append(context);
+
+      BOTAN_ASSERT_NOMSG(prefix_stuffer.full());
```

Comment:
Alternatively (to avoid fiddling with the stuffer and precomputing the required buffer length):

```C++
msg_internal.prefix = concat<SphincsMessagePrefix>(store_be(input_mode_byte),
                                                   store_be(checked_cast_to<uint8_t>(context.size())),
                                                   context);
```

... unfortunately, `concat()` can't deal directly with the atomic underlying type (`uint8_t`), hence the `store_be` invocations (that transform the single byte into a 1-byte array which is a "range"). One day, I'll look into allowing this. :)